ciao-cli/tenant.go

```diff
@@ -17,16 +17,118 @@
 package main
 
 import (
+	"bytes"
+	"encoding/json"
+	"errors"
 	"flag"
 	"fmt"
 	"os"
 	"text/template"
 	"time"
 
+	"github.com/01org/ciao/ciao-controller/api"
 	"github.com/01org/ciao/ciao-controller/types"
 	"github.com/intel/tfortools"
 )
 
+func getCiaoTenantsResource() (string, error) {
+	url, err := getCiaoResource("tenants", api.TenantsV1)
+	return url, err
+}
+
+func getCiaoTenantRef(ID string) (string, error) {
+	var tenants types.TenantsListResponse
+
+	query := queryValue{
+		name:  "id",
+		value: ID,
+	}
+
+	url, err := getCiaoTenantsResource()
+	if err != nil {
+		return "", err
+	}
+
+	if !checkPrivilege() {
+		return url, err
+	}
+
+	resp, err := sendCiaoRequest("GET", url, []queryValue{query}, nil, api.TenantsV1)
+	if err != nil {
+		return "", err
+	}
+
+	err = unmarshalHTTPResponse(resp, &tenants)
+	if err != nil {
+		return "", err
+	}
+
+	if len(tenants.Tenants) != 1 {
+		return "", errors.New("No tenant by that ID found")
+	}
+
+	links := tenants.Tenants[0].Links
+	url = getRef("self", links)
+	if url == "" {
+		return url, errors.New("invalid link returned from controller")
+	}
+
+	return url, nil
+}
+
+func getCiaoTenantConfig(ID string) (types.TenantConfig, error) {
+	var config types.TenantConfig
+
+	url, err := getCiaoTenantRef(ID)
+	if err != nil {
+		return config, err
+	}
+
+	resp, err := sendCiaoRequest("GET", url, nil, nil, api.TenantsV1)
+	if err != nil {
+		return config, err
+	}
+
+	err = unmarshalHTTPResponse(resp, &config)
+
+	return config, err
+}
+
+func putCiaoTenantConfig(ID string, name string, bits int) error {
+	var config types.TenantConfig
+
+	url, err := getCiaoTenantRef(ID)
+	if err != nil {
+		return err
+	}
+
+	resp, err := sendCiaoRequest("GET", url, nil, nil, api.TenantsV1)
```

Comment:
This is a little racy.  If two clients update the tenant information at the same time we could have unexpected results.

If a client does 

ciao-cli tenant update --subnebits

he would expect to just update the subnetbits and not the tenant name.

But if someone else updated the name after this call but before the second call our user makes to sendCiaoRequest, he'd end up inadvertently overwriting their change.

Reply:
@markdryan you are right. This is due to the fact that a PUT is supposed to completely replace the resource. I could change this verb to be a PATCH and that would allow you to only update a single attribute if you wanted to.